gusto/forcing.py

```diff
@@ -26,106 +28,132 @@ def __init__(self, state, euler_poincare=True, linear=False, extra_terms=None):
+        self.sponge = state.mu is not None
+        self.topography = hasattr(state.fields, "topography")
         self.extra_terms = extra_terms
-        self._build_forcing_solver()
+
+        # some constants to use for scaling terms
+        self.scaling = Constant(1.)
+        self.mu_scaling = Constant(1.)
+
+        self._build_forcing_solvers()
+
+    def mass_term(self):
+        return inner(self.test, self.trial)*dx
+
+    def coriolis_term(self):
+        u0 = split(self.x0)[0]
+        return -inner(self.test, cross(2*self.state.Omega, u0))*dx
+
+    def sponge_term(self):
+        u0 = split(self.x0)[0]
+        return self.state.mu*inner(self.test, self.state.k)*inner(u0, self.state.k)*dx
+
+    def euler_poincare_term(self):
+        u0 = split(self.x0)[0]
+        return -0.5*div(self.test)*inner(u0, u0)*dx
 
     @abstractmethod
-    def _build_forcing_solver(self):
+    def pressure_gradient_term(self):
         pass
 
-    @abstractmethod
-    def apply(self, scale, x, x_nl, x_out, **kwargs):
+    def forcing_term(self):
+        L = self.pressure_gradient_term()
+        if self.extruded:
+            L += self.gravity_term()
+        if self.coriolis:
+            L += self.coriolis_term()
+        if self.euler_poincare:
+            L += self.euler_poincare_term()
+        if self.topography:
+            L += self.topography_term()
+        if self.extra_terms is not None:
+            L += inner(self.test, self.extra_terms)*dx
+        # scale L
+        L = self.scaling * L
+        # sponge term has a separate scaling factor as it is always implicit
+        if self.sponge:
+            L += self.mu_scaling*self.sponge_term()
+        return L
+
+    def _build_forcing_solvers(self):
+        a = self.mass_term()
+        L = self.forcing_term()
+        if self.Vu.extruded:
+            bcs = [DirichletBC(self.Vu, 0.0, "bottom"),
+                   DirichletBC(self.Vu, 0.0, "top")]
+        else:
+            bcs = None
+
+        u_forcing_problem = LinearVariationalProblem(
+            a, L, self.uF, bcs=bcs
+        )
+
+        self.u_forcing_solver = LinearVariationalSolver(u_forcing_problem)
+
+    def apply(self, scaling, x_in, x_nl, x_out, **kwargs):
         """
         Function takes x as input, computes F(x_nl) and returns
         x_out = x + scale*F(x_nl)
         as output.
 
-        :arg x: :class:`.Function` object
+        :arg x_in: :class:`.Function` object
         :arg x_nl: :class:`.Function` object
         :arg x_out: :class:`.Function` object
         :arg mu_alpha: scale for sponge term, if present
         """
-        pass
+        self.scaling.assign(scaling)
+        self.x0.assign(x_nl)
+        self.x_out = x_out
+        if 'mu_alpha' in kwargs and kwargs['mu_alpha'] is not None:
+            self.mu_scaling.assign(kwargs['mu_alpha'])
+        self.u_forcing_solver.solve()  # places forcing in self.uF
+
+        uF = self.x_out.split()[0]
```

Comment:
`uF, _` ?  Or is this painful because it could have length 2 or 3 depending on eqn?  (SWE vs compressible Euler, say)